File: admin-web/lib/mesafe.ts
/**
 * (P160) IKI KOORDINAT ARASI MESAFE + ESIK KARARI.
 *
 * =========================================================================
 * ESIK ARTIK VAR — VE BIR AYARDIR
 * =========================================================================
 * Onceki turda burada "sistemde esik YOK, uydurmak urun karari olurdu"
 * yaziyordu. Karar ALINDI (Kerem): varsayilan 50 m, `tenant`ta
 * `okutma_mesafe_esigi_m` olarak tutulur ve Ayarlar'dan degistirilir.
 * Esik SABIT DEGIL cunku site olcekleri cok farkli: bir sitede noktalar
 * 10 m araliklarla, digerinde bloklar arasi 200 m.
 *
 * =========================================================================
 * "BELIRSIZ" UCUNCU BIR SONUCTUR — VE ZORUNLUDUR
 * =========================================================================
 * GPS dogrulugu esikten BUYUKSE karsilastirma KARAR VEREMEZ: ±100 m
 * hatayla olculmus bir mesafenin 50 m esigini gecip gecmedigi
 * bilinemez. Bunu "esik disi" saymak, olcum hatasini ihlal diye
 * raporlamakti — yani birini yanlis suclamak.
 *
 * Bu bir URUN KARARI DEGIL, ARITMETIK: hata payi esigin tamamindan
 * buyukse kiyas anlamsizdir. O yuzden panel ucuncu bir sonuc doner ve
 * mesafeyi yine yazar; yargiyi kullaniciya birakir.
 *
 * Sunucunun kendi gerekcesi de bu yonde (`routers/scans.py`): NTAG424 SDM
 * etiketin FIZIKSEL varligini kriptografik olarak kanitliyor; GPS "konumu
 * ekler", tek basina bir kanit degil.
 *
 * =========================================================================
 * HAVERSINE — neden bu
 * =========================================================================
 * Site olceginde (yuzlerce metre) duzlemsel yaklasim da yeterdi, ama
 * haversine hem kisa hem dogru ve enlem arttikca bozulmuyor. Dunya
 * yaricapi ortalama kure yaricapi (WGS84 ortalama).
 */

/** Ortalama Dunya yaricapi (m) — WGS84 ortalama kure. */
const DUNYA_YARICAPI = 6_371_008.8;

const derece = (d: number) => (d * Math.PI) / 180;

/** Iki nokta arasi buyuk cember mesafesi — METRE, tam sayiya yuvarli. */
export function mesafeMetre(
  aLat: number,
  aLon: number,
  bLat: number,
  bLon: number,
): number {
  const dLat = derece(bLat - aLat);
  const dLon = derece(bLon - aLon);
  const h =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(derece(aLat)) * Math.cos(derece(bLat)) * Math.sin(dLon / 2) ** 2;
  return Math.round(2 * DUNYA_YARICAPI * Math.asin(Math.min(1, Math.sqrt(h))));
}

/** Bir okutmanin esige gore durumu. */
export type EsikSonucu = "icinde" | "disinda" | "belirsiz";

/**
 * Mesafeyi esikle karsilastirir.
 *
 * `dogruluk` bilinmiyorsa (eski istemci) karsilastirma yapilir — elde
 * baska bir sey yok ve "belirsiz" demek her okutmayi belirsiz yapardi.
 */
export function esikSonucu(
  mesafe: number,
  esik: number,
  dogruluk: number | null | undefined,
): EsikSonucu {
  if (dogruluk != null && dogruluk > esik) return "belirsiz";
  return mesafe > esik ? "disinda" : "icinde";
}

File: backend/app/mesafe.py
"""(P160) Okutma mesafesi ve ESIK KARARI — saf hesap, veritabani YOK.

===========================================================================
KURAL ISTEMCIYLE BIREBIR AYNI OLMALI
===========================================================================
Ayni karar iki yerde veriliyor:
  * panel haritasi (`admin-web/lib/mesafe.ts`) — GORSELLESTIRME,
  * burasi — ALARM.

Ikisi ayrisirsa harita "esik icinde" derken bildirim "esik disi" der ve
yonetici hangisine inanacagini bilemez. Bu yuzden kural TEK CUMLEYLE
yazildi ve iki dilde ayni: `dogruluk > esik` ise BELIRSIZ, degilse
`mesafe > esik` ise DISINDA, aksi halde ICINDE.

===========================================================================
"BELIRSIZ" NEDEN ZORUNLU
===========================================================================
GPS dogrulugu esikten BUYUKSE karsilastirma KARAR VEREMEZ: ±100 m hatayla
olculmus bir mesafenin 50 m esigini gecip gecmedigi bilinemez. Bunu "esik
disi" saymak, OLCUM HATASINI IHLAL diye raporlamakti — ve alarm uretmek,
haritada renk degistirmekten cok daha agir bir iddiadir: birinin
telefonu caliyor.

Bu bir urun karari DEGIL, ARITMETIK: hata payi esigin tamamindan buyukse
kiyas anlamsizdir. BELIRSIZ olan okutma icin ALARM URETILMEZ.

===========================================================================
HAVERSINE
===========================================================================
Site olceginde duzlemsel yaklasim da yeterdi ama haversine hem kisa hem
dogru ve enlem arttikca bozulmuyor. Yaricap WGS84 ortalama kure yaricapi
— `admin-web/lib/mesafe.ts` ile AYNI sabit.
"""
from __future__ import annotations

from math import asin, cos, radians, sin, sqrt
from typing import Literal

#: Ortalama Dunya yaricapi (m) — WGS84 ortalama kure.
DUNYA_YARICAPI = 6_371_008.8

EsikSonucu = Literal["icinde", "disinda", "belirsiz"]


def mesafe_metre(a_lat: float, a_lon: float, b_lat: float, b_lon: float) -> int:
    """Iki nokta arasi buyuk cember mesafesi — METRE, tam sayiya yuvarli."""
    d_lat = radians(b_lat - a_lat)
    d_lon = radians(b_lon - a_lon)
    h = (
        sin(d_lat / 2) ** 2
        + cos(radians(a_lat)) * cos(radians(b_lat)) * sin(d_lon / 2) ** 2
    )
    return round(2 * DUNYA_YARICAPI * asin(min(1.0, sqrt(h))))


def esik_sonucu(mesafe: int, esik: int, dogruluk: float | None) -> EsikSonucu:
    """Mesafeyi esikle karsilastirir — `mesafe.ts`teki kuralin aynisi.

    `dogruluk` bilinmiyorsa (eski istemci alani gondermez) karsilastirma
    YAPILIR: elde baska bir sey yok ve her okutmayi belirsiz saymak esigi
    tamamen ise yaramaz kilardi.
    """
    if dogruluk is not None and dogruluk > esik:
        return "belirsiz"
    return "disinda" if mesafe > esik else "icinde"
